cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -113,27 +113,55 @@ func New(endpoints []string, ca, cert, key string) (*Client, error) {
 	return &client, nil
 }
 
-// NewFromStaticPod creates a GenericClient from the given endpoints, manifestDir, and certificatesDir
-func NewFromStaticPod(endpoints []string, manifestDir string, certificatesDir string) (*Client, error) {
-	hasTLS, err := PodManifestsHaveTLS(manifestDir)
+// NewFromCluster creates an etcd client for the the etcd endpoints defined in the ClusterStatus value stored in
+// the kubeadm-config ConfigMap in kube-system namespace.
+// Once created, the client synchronizes client's endpoints with the known endpoints from the etcd membership API (reality check).
+func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client, error) {
+
+	// Kubeadm v1.13 should manage v1.12 clusters and v1.13 clusters
+	// v1.12 clusters can be have etcd listening on localhost only (if the cluster was created with kubeadm v1.12)
+	// or etcd listening on localhost and API server advertise address (if the cluster was created with kubeadm v1.13).
+	// The first case should be dropped in v1.14 when support for v1.12 clusters can be removed from the codebase.
+
+	// Detect which type of etcd we are dealing with
+	oldManifest := false
+	glog.V(1).Infoln("checking etcd manifest")
+
+	etcdManifestFile := constants.GetStaticPodFilepath(constants.Etcd, constants.GetStaticPodDirectory())
+	etcdPod, err := staticpod.ReadStaticPodFromDisk(etcdManifestFile)
 	if err != nil {
-		return nil, errors.Wrapf(err, "could not read manifests from: %s, error", manifestDir)
+		return nil, errors.Wrap(err, "error reading etcd manifest file")
+	}
+	etcdContainer := etcdPod.Spec.Containers[0]
+	for _, arg := range etcdContainer.Command {
+		if arg == "--listen-client-urls=https://127.0.0.1:2379" {
```

Comment:
i guess there is no other way to determine if these pod manifests were created by kubeadm 1.12?

Reply:
I'm not a fan of this solution, but considering it should go aways next cycle I think this can be accepted (the same hack is used elsewhere in kubeadm itself).
Nevertheless, if there are any better idea ...

Reply:
we probably should just merge on monday unless there is further feedback.

Reply:
Works for me.
In the future it may be handy to add a comment at the beginning of generated YAML files, stating that these files are autogenerated and the kubeadm version with which it was done. This will not only provide a start point for issues like this, but make users less tempted to tamper the generated YAMLs themselves, rather than edit the kubeadm config and rerun it.

Reply:
Very good point @rosti, please file an issue for it!

Reply:
@luxas @rosti what about adding an annotation too? comments are fine, but cannot be used by app

Reply:
for 1.12 -> 1.13 there is no way.

for later versions an annotation seem OK as long as we need it and/or want to maintain it..
annotating which pods are created with what kubeadm version seems like a nice to have to me.

```
kind: Pod
metadata:
  annotations:
    kubeadmVersion: <version-from-client>
```
`version-from-client` should be the same as `kubeadm version -o=short`
and be parsed using: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/version/version.go#L111

Reply:
issue created:
https://github.com/kubernetes/kubeadm/issues/1231